### ISSUE TITLE
Bean Validation composed constraints should be supported

### ISSUE THREAD

**Issue by astiob** [typetools/checker-framework#1668]
This is similar to part of #1665, so maybe you’ll reject this as well, but here goes…

Bean Validation (version 1 and 2 alike) [allows](http://beanvalidation.org/2.0/spec/#constraintsdefinitionimplementation-constraintcomposition) annotating a constraint with another constraint to inherit its meaning. For example, if you write

```java
@Target({ElementType.METHOD, ElementType.FIELD, ElementType.ANNOTATION_TYPE,
         ElementType.CONSTRUCTOR, ElementType.PARAMETER, ElementType.TYPE_USE})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy={})
@NotNull
@interface MyNotNull {
    String message() default "";
    Class<?>[] groups() default {};
    Class<? extends Payload>[] payload() default {};
}
```

then you can use `@MyNotNull` as an alias for `@NotNull`.

Checker Framework doesn’t understand this and ignores the custom constraint.

Why is this useful? Well, in my case, I was using a validator that supported `TYPE_USE` constraints in accordance with the Bean Validation 2.0 spec but brought in an old Bean Validation 1.1 API dependency that didn’t declare `TYPE_USE`. I didn’t want to risk breaking something else by including the whole newer API, so I wrote code much like the above and used the custom constraint in my validated beans.

Another reason might be to combine several constraints into one, if I understand correctly that this is possible. It might not seem to matter much for Checker Framework as `NotNull` is the only Bean Validation constraint that Checker Framework supports, but if the programmer combines `NotNull` with some other constraints and uses the new constraint, Checker Framework should still pick up the `NotNull`.

By the way, Checker Framework currently (2.2.2) ships with an outdated definitions of `javax.validation.constraints.NotNull` that doesn’t declare `TYPE_USE`. Now that I’ve upgraded to a newer validator and no longer need my custom constraint, I’ve discovered that Checker Framework’s outdated definition overrides the newer definition from the validator and I have to tweak my dependencies to override it again. Does Checker Framework have to ship all those third-party annotations at all?

**Comment by wmdietl** [comment]
As mentioned in #1664 there currently is only a programmatic way to add aliases for type qualifiers.
It would be a welcome contribution to add a more flexible aliasing mechanism, which should then be usable for these custom composed constraints, too.

Thanks for pointing out that Java EE 8 added `TYPE_USE` to some of those annotations, as documented here https://javaee.github.io/javaee-spec/javadocs/javax/validation/constraints/NotNull.html
Do you want to work on a PR that updates the relevant annotations?

**Comment by mernst** [comment]
I've fixed the `javax.validation.constraints.NotNull` annotation in commit deb90b115.  My apologies for the bug, and thank you for pointing it out.

We ought to periodically check all the external annotations to see whether they have been updated.  (Though I don't think I would have noticed it in this case -- a Google search for "javax.validation.constraints.NotNull" yielded https://docs.oracle.com/javaee/7/api/javax/validation/constraints/NotNull.html as the second hit, and didn't have https://javaee.github.io/javaee-spec/javadocs/javax/validation/constraints/NotNull.html in the top 100 hits.)  If you notice others that are out of date, please let us know.

**Comment by astiob** [comment]
Thanks, folks! No worries—you can’t keep tabs on everything at once.

Is there a particular reason why Checker Framework ships third-party annotations in the first place? If it is just to make Nullness Checker itself compile, it probably should be possible to look them up via reflection and silently skip any that are missing.

**Comment by mernst** [comment]
I don't remember exactly, but here are some possibilities for why clean-room implementations of the annotations appear in the Checker Framework distribution, rather than using a reflection-like mechanism.

 * We thought that new aliases would arise infrequently, so it didn't seem worthwhile to write and test a more elaborate mechanism.
 * The JVM ignores annotations that are not on the classpath.  Therefore, it's possible that the Checker Framework would ignore the annotation, which would be very surprising to users and very difficult for the Checker Framework developers to debug remotely.  Users would say, "The classfile contains `@Nullable` and the JVM runs, but the Checker Framework is ignoring the annotation."  Ensuring that the annotations are on the classpath removes this possible source of errors and confusion.
 * We wished the annotations to apply to the correct locations (such as `TYPE_USE').

You have brought up a counterexample to the first point.
Bean validation composed constraints are new aliases that may be company-specific or project-specific.

If the second point is not a real problem or can be worked around, then changing to the reflective mechanism would support the composed-constraint use case.

**Comment by wmdietl** [comment]
Just a comment about the last point: I think we could do both, keep the existing definitions for the standard annotations and add a command-line option to provide reflectively-loaded additional aliases.
Something like: `-Aextraaliases=o.c.c.n.Nullable:package.MyNullable` with some convention for encoding multiple aliases.

**Comment by mernst** [comment]
Here are two implementation strategies:

 * via meta-annotations.  The Checker Framework must define a new meta-annotation (or know about some existing meta-annotation defined by another project).  The user must meta-annotate the aliases.  The Checker Framework must  examine all annotations on the classpath and determine whether they are meta-annotated to make them aliases.  This will probably make Checker Framework startup slow.
 * via command-line options.  This requires the user to supply the command-line arguments when invoking the Checker Framework.  These probably will not change very often.  This is simpler to implement.

**Comment by vlsi** [comment]
What do you think of adding an option like `annotation X is exactly the same as Y except package name"?
